hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/ConfiguredFailoverProxyProvider.java

```diff
@@ -93,4 +109,65 @@ public synchronized void close() throws IOException {
   public boolean useLogicalURI() {
     return true;
   }
+
+  /**
+   * Write active NameNode index to cache file.
+   */
+  public void writeActiveCache(int index) {
+    if (!cacheActiveEnabled) {
+      return;
+    }
+
+    synchronized (cacheActiveFile) {
+      boolean exist = cacheActiveFile.exists();
+      try (RandomAccessFile raf = new RandomAccessFile(cacheActiveFile, "rw");
+          FileChannel fc = raf.getChannel();
+          FileLock lock = fc.tryLock(0, Long.MAX_VALUE, false)) {
+        if (lock != null) {
+          raf.setLength(0);
+          raf.writeBytes(String.valueOf(index));
+          if (!exist) {
+            boolean ret = cacheActiveFile.setWritable(true, false)
+                && cacheActiveFile.setReadable(true, false);
+            if (!ret) {
+              throw new IOException("Cannot set file rw mode.");
+            }
+          }
+          LOG.debug("Succeed in writing active index " + index
+              + " to cache file " + cacheActiveFile);
+        }
+      } catch (Throwable e) {
+        LOG.warn("Filed to write active index " + index + " to cache file "
+            + cacheActiveFile, e);
+      }
+    }
+  }
+
+  /**
+   * Read active NameNode index from cache file.
+   */
+  public int readActiveCache() {
+    if (!cacheActiveEnabled) {
+      return 0;
+    }
+
+    int index = 0;
+    synchronized (cacheActiveFile) {
+      if (!cacheActiveFile.exists()) {
+        return 0;
+      }
+
+      try (RandomAccessFile raf = new RandomAccessFile(cacheActiveFile, "rw");
+          FileChannel fc = raf.getChannel();
+          FileLock lock = fc.tryLock(0, Long.MAX_VALUE, true)) {
+        if (lock != null) {
+          index = Integer.parseInt(raf.readLine()) % proxies.size();
```

Comment:
can we double check here to make sure the index is non-negative?

Reply:
OK, I will add the double check.

Reply:
Sorry, I have missed this comment before,  Done now.